packages/plugin-pnpm/sources/PnpmLinker.ts

```diff
@@ -135,25 +148,51 @@ class PnpmInstaller implements Installer {
     if (!isPnpmVirtualCompatible(locator, {project: this.opts.project}))
       return;
 
+    const storeLocation = getStoreLocation(this.opts.project);
+
     this.asyncActions.reduce(locator.locatorHash, async action => {
       // Wait that the package is properly installed before starting to copy things into it
       await action;
 
-      const pkgPath = this.packageLocations.get(locator.locatorHash);
+      const pkgPath = this.customData.packageLocations.get(locator.locatorHash);
       if (typeof pkgPath === `undefined`)
         throw new Error(`Assertion failed: Expected the package to have been registered (${structUtils.stringifyLocator(locator)})`);
 
-      const nmPath = ppath.join(pkgPath, Filename.nodeModules);
-      if (dependencies.length > 0)
-        await xfs.mkdirpPromise(nmPath);
+      const locatorName = structUtils.stringifyIdent(locator) as PortablePath;
+
+      let nmPath: PortablePath;
+      let storeEntryPathToClean: PortablePath | null = null;
+
+      if (ppath.contains(storeLocation, pkgPath) && pkgPath.endsWith(structUtils.getIdentVendorPath(locator))) {
+        nmPath = pkgPath.slice(0, -(ppath.sep.length + locatorName.length)) as PortablePath;
+        storeEntryPathToClean = ppath.dirname(nmPath);
+      } else {
+        nmPath = ppath.join(pkgPath, Filename.nodeModules);
+      }
```

Comment:
I don't understand this condition

Reply:
Removed this part in https://github.com/yarnpkg/berry/pull/3681/commits/a09b37ad067db44e62ede75061b4694319d11263.